File: routes/fields.py
from flask import Blueprint, render_template, request, jsonify
from datetime import datetime
from db_helpers import query_db, modify_db

# Create a Blueprint for fields routes
fields_bp = Blueprint('fields', __name__)

@fields_bp.route('/')
def fields_page():
    return render_template('fields.html', active_page='fields')

@fields_bp.route('/get_field_classes')
def get_field_classes():
    classes = query_db('SELECT * FROM GEE_FIELD_CLASSES')
    return jsonify([dict(cls) for cls in classes])

@fields_bp.route('/get_fields')
def get_fields():
    page = request.args.get('page', 1, type=int)
    per_page = request.args.get('per_page', 10, type=int)
    search = request.args.get('search', '', type=str)
    class_filter = request.args.get('class_filter', '', type=str)
    
    # Base query for counting total records
    count_query = '''
        SELECT COUNT(*) as total
        FROM GEE_FIELDS f 
        LEFT JOIN GEE_FIELD_CLASSES fc ON f.GFC_ID = fc.GFC_ID
    '''
    
    # Base query for fetching records
    base_query = '''
        SELECT f.*, fc.FIELD_CLASS_NAME, fc.CLASS_TYPE
        FROM GEE_FIELDS f 
        LEFT JOIN GEE_FIELD_CLASSES fc ON f.GFC_ID = fc.GFC_ID
    '''
    
    # Build WHERE clause and parameters
    params = []
    where_conditions = []
    
    if search:
        where_conditions.append('(f.GF_NAME LIKE ? OR f.GF_TYPE LIKE ? OR f.GF_DESCRIPTION LIKE ? OR fc.FIELD_CLASS_NAME LIKE ?)')
        search_param = f'%{search}%'
        params.extend([search_param, search_param, search_param, search_param])
    
    if class_filter:
        where_conditions.append('f.GFC_ID = ?')
        params.append(class_filter)
    
    where_clause = ' WHERE ' + ' AND '.join(where_conditions) if where_conditions else ''
    
    # Get total count
    total_query = count_query + where_clause
    total_result = query_db(total_query, params, one=True)
    total = total_result['total'] if total_result else 0
    
    # Calculate pagination
    offset = (page - 1) * per_page
    total_pages = (total + per_page - 1) // per_page
    
    # Fetch paginated data
    data_query = base_query + where_clause + '''
        ORDER BY fc.FIELD_CLASS_NAME, f.GF_NAME
        LIMIT ? OFFSET ?
    '''
    
    fields = query_db(data_query, params + [per_page, offset])
    
    return jsonify({
        'data': [dict(field) for field in fields],
        'pagination': {
            'page': page,
            'per_page': per_page,
            'total': total,
            'total_pages': total_pages,
            'has_prev': page > 1,
            'has_next': page < total_pages
        }
    })

@fields_bp.route('/get_fields_by_class/<int:class_id>')
def get_fields_by_class(class_id):
    fields = query_db('SELECT * FROM GEE_FIELDS WHERE GFC_ID = ?', (class_id,))
    return jsonify([dict(field) for field in fields])

@fields_bp.route('/get_child_classes/<int:parent_class_id>')
def get_child_classes(parent_class_id):
    child_classes = query_db('SELECT * FROM GEE_FIELD_CLASSES WHERE PARENT_GFC_ID = ?', (parent_class_id,))
    return jsonify([dict(cls) for cls in child_classes])

@fields_bp.route('/add_field', methods=['POST'])
def add_field():
    data = request.json
    try:
        modify_db(
            'INSERT INTO GEE_FIELDS (GFC_ID, GF_NAME, GF_TYPE, GF_SIZE, GF_PRECISION_SIZE, GF_DEFAULT_VALUE, GF_DESCRIPTION) VALUES (?, ?, ?, ?, ?, ?, ?)',
            (data['gfcId'], data['fieldName'], data['type'], data['size'], data['precision'], data['defaultValue'], data['description'])
        )
        return jsonify({'success': True, 'message': 'Field added successfully!'})
    except Exception as e:
        return jsonify({'success': False, 'message': str(e)})

@fields_bp.route('/update_field', methods=['PUT'])
def update_field():
    data = request.json
    try:
        modify_db(
            'UPDATE GEE_FIELDS SET GFC_ID = ?, GF_NAME = ?, GF_TYPE = ?, GF_SIZE = ?, GF_PRECISION_SIZE = ?, GF_DEFAULT_VALUE = ?, GF_DESCRIPTION = ?, UPDATE_DATE = ? WHERE GF_ID = ?',
            (data['gfcId'], data['fieldName'], data['type'], data['size'], data['precision'], data['defaultValue'], data['description'], datetime.now(), data['gfId'])
        )
        return jsonify({'success': True, 'message': 'Field updated successfully!'})
    except Exception as e:
        return jsonify({'success': False, 'message': str(e)})

@fields_bp.route('/delete_field/<int:gf_id>', methods=['DELETE'])
def delete_field(gf_id):
    try:
        modify_db('DELETE FROM GEE_FIELDS WHERE GF_ID = ?', (gf_id,))
        return jsonify({'success': True, 'message': 'Field deleted successfully!'})
    except Exception as e:
        return jsonify({'success': False, 'message': str(e)})

@fields_bp.route('/bulk_delete_fields', methods=['DELETE'])
def bulk_delete_fields():
    """Delete multiple fields"""
    try:
        data = request.json
        field_ids = data.get('field_ids', [])
        
        if not field_ids:
            return jsonify({'success': False, 'message': 'No field IDs provided'})
        
        deleted_count = 0
        errors = []
        
        for field_id in field_ids:
            try:
                # Check if field exists
                existing_field = query_db('SELECT GF_NAME FROM GEE_FIELDS WHERE GF_ID = ?', (field_id,), one=True)
                if not existing_field:
                    continue
                
                # Delete the field
                modify_db('DELETE FROM GEE_FIELDS WHERE GF_ID = ?', (field_id,))
                deleted_count += 1
                
            except Exception as e:
                errors.append(f"Error deleting field {field_id}: {str(e)}")
        
        message = f'Bulk delete completed. Deleted: {deleted_count} field(s)'
        if errors:
            message += f'. Errors: {len(errors)}'
        
        return jsonify({
            'success': True, 
            'message': message,
            'deleted_count': deleted_count,
            'errors': errors
        })
        
    except Exception as e:
        return jsonify({'success': False, 'message': f'Bulk delete failed: {str(e)}'})

File: static/js/fields.js
// Global variables for field management
let currentDeleteId = null;
let fieldModal = null;
let deleteModal = null;
let currentFieldPage = 1;
let currentFieldSearch = '';
let currentFieldClassFilter = '';

// Initialize when DOM is loaded
document.addEventListener('DOMContentLoaded', function() {
    // Initialize Bootstrap modals
    const fieldModalElement = document.getElementById('fieldModal');
    if (fieldModalElement) {
        fieldModal = new bootstrap.Modal(fieldModalElement);
    }
    
    const deleteModalElement = document.getElementById('deleteModal');
    if (deleteModalElement) {
        deleteModal = new bootstrap.Modal(deleteModalElement);
    }
    
    // Load initial data
    loadFields();
    loadFieldClasses();
    loadFilterClasses();
});

// Load field classes for the form dropdown
function loadFieldClasses() {
    fetch('/fields/get_field_classes')
        .then(response => response.json())
        .then(data => {
            const select = document.getElementById('fieldClass');
            select.innerHTML = '<option value="">Select Class</option>';
            data.forEach(cls => {
                select.innerHTML += `<option value="${cls.GFC_ID}">${cls.FIELD_CLASS_NAME}</option>`;
            });
        })
        .catch(error => {
            console.error('Error loading field classes:', error);
            showToast('Error', 'Failed to load field classes', 'error');
        });
}

// Load field classes for the filter dropdown
function loadFilterClasses() {
    fetch('/fields/get_field_classes')
        .then(response => response.json())
        .then(data => {
            const select = document.getElementById('filterFieldClass');
            select.innerHTML = '<option value="">All Classes</option>';
            data.forEach(cls => {
                select.innerHTML += `<option value="${cls.GFC_ID}">${cls.FIELD_CLASS_NAME}</option>`;
            });
        })
        .catch(error => {
            console.error('Error loading filter classes:', error);
        });
}

// Load all fields
function loadFields(page = 1, search = '', classFilter = '') {
    currentFieldPage = page;
    currentFieldSearch = search;
    currentFieldClassFilter = classFilter;
    
    let url = `/fields/get_fields?page=${page}&per_page=10`;
    if (search) {
        url += `&search=${encodeURIComponent(search)}`;
    }
    if (classFilter) {
        url += `&class_filter=${encodeURIComponent(classFilter)}`;
    }
    
    fetch(url)
        .then(response => response.json())
        .then(response => {
            const data = response.data;
            const pagination = response.pagination;
            
            renderFields(data);
            updateFieldPaginationControls(pagination);
        })
        .catch(error => {
            console.error('Error loading fields:', error);
            showToast('Error', 'Failed to load fields', 'error');
        });
}

// Render fields to the table
function renderFields(fields) {
    const tableBody = document.getElementById('fieldTableBody');
    tableBody.innerHTML = '';
    
    fields.forEach(field => {
        const className = field.FIELD_CLASS_NAME || 'No Class';
        const row = `
            <tr>
                <td>
                    <input type="checkbox" class="field-checkbox" value="${field.GF_ID}" onchange="updateBulkDeleteFieldsButton()">
                </td>
                <td>${field.GF_NAME}</td>
                <td>
                    <span class="badge bg-secondary">${className}</span>
                </td>
                <td>${field.GF_TYPE}</td>
                <td>${field.GF_SIZE || ''}</td>
                <td>${field.GF_PRECISION_SIZE || ''}</td>
                <td>${field.GF_DEFAULT_VALUE || ''}</td>
                <td>${field.GF_DESCRIPTION || ''}</td>
                <td class="action-buttons">
                    <button class="btn btn-warning btn-sm" onclick="editField(${JSON.stringify(field).replace(/"/g, '&quot;')})">
                        <i class="fas fa-edit"></i>
                    </button>
                    <button class="btn btn-danger btn-sm" onclick="deleteField(${field.GF_ID})">
                        <i class="fas fa-trash"></i>
                    </button>
                </td>
            </tr>
        `;
        tableBody.innerHTML += row;
    });

    // Update field count
    document.getElementById('fieldCount').textContent = `${fields.length} fields`;
}

// Open modal for creating new field
function openNewFieldModal() {
    document.getElementById('modalTitle').textContent = 'Add New Field';
    document.getElementById('fieldForm').reset();
    document.getElementById('gfId').value = '';
    fieldModal.show();
}

// Open modal for editing existing field
function editField(field) {
    document.getElementById('modalTitle').textContent = 'Edit Field';
    document.getElementById('gfId').value = field.GF_ID;
    document.getElementById('fieldClass').value = field.GFC_ID || '';
    document.getElementById('fieldName').value = field.GF_NAME;
    document.getElementById('fieldType').value = field.GF_TYPE;
    document.getElementById('fieldSize').value = field.GF_SIZE || '';
    document.getElementById('fieldPrecision').value = field.GF_PRECISION_SIZE || '';
    document.getElementById('fieldDefault').value = field.GF_DEFAULT_VALUE || '';
    document.getElementById('fieldDescription').value = field.GF_DESCRIPTION || '';
    fieldModal.show();
}

// Save field (create or update)
function saveField() {
    const gfId = document.getElementById('gfId').value;
    const data = {
        gfcId: document.getElementById('fieldClass').value || null,
        fieldName: document.getElementById('fieldName').value,
        type: document.getElementById('fieldType').value,
        size: document.getElementById('fieldSize').value || null,
        precision: document.getElementById('fieldPrecision').value || null,
        defaultValue: document.getElementById('fieldDefault').value || null,
        description: document.getElementById('fieldDescription').value || null
    };

    if (!data.fieldName || !data.type) {
        showToast('Error', 'Field Name and Type are required!', 'error');
        return;
    }

    const method = gfId ? 'PUT' : 'POST';
    const url = gfId ? '/fields/update_field' : '/fields/add_field';
    if (gfId) data.gfId = gfId;

    fetch(url, {
        method: method,
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify(data)
    })
    .then(response => response.json())
    .then(result => {
        if (result.success) {
            fieldModal.hide();
            loadFields();
            showToast('Success', result.message);
        } else {
            showToast('Error', result.message, 'error');
        }
    })
    .catch(error => {
        console.error('Error saving field:', error);
        showToast('Error', 'Failed to save field', 'error');
    });
}

// Filter fields based on search criteria
function filterFields() {
    const searchTerm = document.getElementById('searchFields').value.toLowerCase();
    const classFilter = document.getElementById('filterFieldClass').value;
    const typeFilter = document.getElementById('filterFieldType').value;
    
    let filteredFields = allFields;
    
    // Apply search filter
    if (searchTerm) {
        filteredFields = filteredFields.filter(field => 
            field.GF_NAME.toLowerCase().includes(searchTerm) || 
            (field.GF_DESCRIPTION && field.GF_DESCRIPTION.toLowerCase().includes(searchTerm)) ||
            (field.FIELD_CLASS_NAME && field.FIELD_CLASS_NAME.toLowerCase().includes(searchTerm))
        );
    }
    
    // Apply class filter
    if (classFilter) {
        filteredFields = filteredFields.filter(field => 
            field.GFC_ID && field.GFC_ID.toString() === classFilter
        );
    }
    
    // Apply type filter
    if (typeFilter) {
        filteredFields = filteredFields.filter(field => field.GF_TYPE === typeFilter);
    }
    
    renderFields(filteredFields);
    
    // Reset checkbox states after filtering
    document.getElementById('selectAllFields').checked = false;
    updateBulkDeleteFieldsButton();
}

// Delete field
function deleteField(gfId) {
    currentDeleteId = gfId;
    deleteModal.show();
}

// Confirm deletion
function confirmDelete() {
    if (!currentDeleteId) return;
    
    fetch(`/fields/delete_field/${currentDeleteId}`, {
        method: 'DELETE'
    })
    .then(response => response.json())
    .then(data => {
        if (data.success) {
            deleteModal.hide();
            loadFields();
            showToast('Success', data.message);
        } else {
            showToast('Error', data.message, 'error');
        }
    })
    .catch(error => {
        console.error('Error deleting field:', error);
        showToast('Error', 'Failed to delete field', 'error');
    });
}

// ======================================
// BULK DELETE FUNCTIONALITY
// ======================================

// Toggle all field checkboxes
function toggleAllFields(selectAllCheckbox) {
    const checkboxes = document.querySelectorAll('.field-checkbox');
    checkboxes.forEach(checkbox => {
        checkbox.checked = selectAllCheckbox.checked;
    });
    updateBulkDeleteFieldsButton();
}

// Update bulk delete button based on selected checkboxes
function updateBulkDeleteFieldsButton() {
    const selectedCheckboxes = document.querySelectorAll('.field-checkbox:checked');
    const bulkDeleteBtn = document.getElementById('bulkDeleteFieldsBtn');
    const selectAllCheckbox = document.getElementById('selectAllFields');
    
    if (selectedCheckboxes.length > 0) {
        bulkDeleteBtn.style.display = 'inline-block';
        bulkDeleteBtn.innerHTML = `<i class="fas fa-trash me-2"></i>Delete Selected (${selectedCheckboxes.length})`;
    } else {
        bulkDeleteBtn.style.display = 'none';
    }

    // Update select all checkbox state
    const allCheckboxes = document.querySelectorAll('.field-checkbox');
    if (selectedCheckboxes.length === allCheckboxes.length) {
        selectAllCheckbox.checked = true;
        selectAllCheckbox.indeterminate = false;
    } else if (selectedCheckboxes.length > 0) {
        selectAllCheckbox.checked = false;
        selectAllCheckbox.indeterminate = true;
    } else {
        selectAllCheckbox.checked = false;
        selectAllCheckbox.indeterminate = false;
    }
}

// Bulk delete selected fields
function bulkDeleteFields() {
    const selectedCheckboxes = document.querySelectorAll('.field-checkbox:checked');
    if (selectedCheckboxes.length === 0) {
        showToast('Error', 'Please select fields to delete', 'error');
        return;
    }

    const selectedIds = Array.from(selectedCheckboxes).map(cb => cb.value);
    const confirmMessage = `Are you sure you want to delete ${selectedIds.length} selected field(s)? This action cannot be undone.`;
    
    if (confirm(confirmMessage)) {
        fetch('/fields/bulk_delete_fields', {
            method: 'DELETE',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                field_ids: selectedIds
            })
        })
        .then(response => response.json())
        .then(data => {
            if (data.success) {
                loadFields();
                document.getElementById('selectAllFields').checked = false;
                updateBulkDeleteFieldsButton();
                showToast('Success', `Successfully deleted ${data.deleted_count} field(s).`);
            } else {
                showToast('Error', data.message, 'error');
            }
        })
        .catch(error => {
            console.error('Error:', error);
            showToast('Error', 'Error during bulk delete operation', 'error');
        });
    }
}

// ======================================
// UTILITY FUNCTIONS
// ======================================

// Show toast notification
function showToast(title, message, type = 'success') {
    // Use global showToast if available
    if (typeof window.showToast === 'function') {
        window.showToast(title, message, type);
        return;
    }
    
    // Fallback alert
    alert(`${title}: ${message}`);
}

// ======================================
// PAGINATION FUNCTIONS
// ======================================

function updateFieldPaginationControls(pagination) {
    const paginationContainer = document.getElementById('fieldPagination');
    if (!paginationContainer) return;
    
    const { page, per_page, total, total_pages, has_prev, has_next } = pagination;
    
    // Update pagination info
    const paginationInfo = document.getElementById('fieldPaginationInfo');
    if (paginationInfo) {
        const start = (page - 1) * per_page + 1;
        const end = Math.min(page * per_page, total);
        paginationInfo.textContent = `Showing ${start}-${end} of ${total} fields`;
    }
    
    let paginationHtml = '<nav><ul class="pagination justify-content-center">';
    
    // Previous button
    paginationHtml += `
        <li class="page-item ${!has_prev ? 'disabled' : ''}">
            <a class="page-link" href="#" onclick="${has_prev ? `changeFieldPage(${page - 1})` : 'return false;'}">
                <i class="fas fa-chevron-left"></i> Previous
            </a>
        </li>
    `;
    
    // Page numbers
    const startPage = Math.max(1, page - 2);
    const endPage = Math.min(total_pages, page + 2);
    
    if (startPage > 1) {
        paginationHtml += `
            <li class="page-item">
                <a class="page-link" href="#" onclick="changeFieldPage(1)">1</a>
            </li>
        `;
        if (startPage > 2) {
            paginationHtml += '<li class="page-item disabled"><span class="page-link">...</span></li>';
        }
    }
    
    for (let i = startPage; i <= endPage; i++) {
        paginationHtml += `
            <li class="page-item ${i === page ? 'active' : ''}">
                <a class="page-link" href="#" onclick="changeFieldPage(${i})">${i}</a>
            </li>
        `;
    }
    
    if (endPage < total_pages) {
        if (endPage < total_pages - 1) {
            paginationHtml += '<li class="page-item disabled"><span class="page-link">...</span></li>';
        }
        paginationHtml += `
            <li class="page-item">
                <a class="page-link" href="#" onclick="changeFieldPage(${total_pages})">${total_pages}</a>
            </li>
        `;
    }
    
    // Next button
    paginationHtml += `
        <li class="page-item ${!has_next ? 'disabled' : ''}">
            <a class="page-link" href="#" onclick="${has_next ? `changeFieldPage(${page + 1})` : 'return false;'}">
                Next <i class="fas fa-chevron-right"></i>
            </a>
        </li>
    `;
    
    paginationHtml += '</ul></nav>';
    paginationContainer.innerHTML = paginationHtml;
}

function changeFieldPage(page) {
    loadFields(page, currentFieldSearch, currentFieldClassFilter);
}

function performFieldSearch() {
    const searchInput = document.getElementById('fieldSearch');
    const classFilter = document.getElementById('filterFieldClass');
    const searchTerm = searchInput ? searchInput.value.trim() : '';
    const selectedClass = classFilter ? classFilter.value : '';
    loadFields(1, searchTerm, selectedClass);
}

function clearFieldSearch() {
    const searchInput = document.getElementById('fieldSearch');
    const classFilter = document.getElementById('filterFieldClass');
    if (searchInput) {
        searchInput.value = '';
    }
    if (classFilter) {
        classFilter.value = '';
    }
    loadFields(1, '', '');
}